### ISSUE TITLE
BUG: Back test, analyse.py strips last candle as incomplete, in BT its never incomplete.

### ISSUE THREAD

**Issue by creslinux** [freqtrade/freqtrade#1011]
In Backtest analyse.py is striping the last candle of the dataframe as incomplete. 
No candle data in backtest is incomplete so this is bug behaviour. 

`  frame.drop(frame.tail(1).index, inplace=True)     # eliminate partial candle`

This effectively means we've triggered our buy/sell/or stop on current bar which we then throw away before calculating returns on purchase/sale.

The leads to false (past) pricing presented when opening/closing a trade

As example impact to results with last candle stripped and without. 
Same data/strategy etc.

> ======================================== BACKTESTING REPORT =========================================
> | pair     |   buy count |   avg profit % |   total profit USDT |   avg duration |   profit |   loss |
> |:---------|------------:|---------------:|--------------------:|---------------:|---------:|-------:|
> | BTC/USDT |           1 |          -0.20 |         -0.00203731 |          360.0 |        0 |      1 |
> | TOTAL    |           1 |          -0.20 |         -0.00203731 |          360.0 |        0 |      1 |
> 

And with the guilty line hashed out: 

> 2018-07-10 15:08:43,341 - freqtrade.optimize.backtesting - INFO - 
> ======================================== BACKTESTING REPORT =========================================
> | pair     |   buy count |   avg profit % |   total profit USDT |   avg duration |   profit |   loss |
> |:---------|------------:|---------------:|--------------------:|---------------:|---------:|-------:|
> | BTC/USDT |           1 |          -6.92 |         -0.06930524 |          480.0 |        0 |      1 |
> | TOTAL    |           1 |          -6.92 |         -0.06930524 |          480.0 |        0 |      1 |

**Comment by creslinux** [comment]
As some exchanges do and some do not present incomplete candles and back test we cannot use time to test i would propose this be fixed more completely to save it rearing up later as we deal with different exchange behaviours 

1) if backtest - do not strip last row
2) if not backtest
     - check if candle date is less than ticker_interval/timeframe in the past
     - - True discard
     - - False keep 

**### The logic being:**
**In Backtest**
We cannot test for time, but know candles are complete so the last row should not be removed.

**In DryTest / Live**
All candles use their opening time as the timestamp. 
So a 1hr candle less than 1hour old is incomplete and should be discarded.
A 1hour candle more than 1 hour old is has completed and can be used.

**Comment by xmatthias** [comment]
i disagree that in backtest candles are always complete.
try using the download-script for a longer timeframe (say 4h) and you will notice that the last candle is not complete (at least for binance). 

Backtest data download uses the same exact function to get data than real trading, so is also impacted by the same issues.

However, i agree that we could (and probably should) change `get_ticker_history` to only return guaranteed full candles (`open-time < utcnow - ticker_timeframe`)

**Comment by creslinux** [comment]
In backtest a block of candles are downloaded and proccessed as a single block
They are then after strat presented to analyse 

If there are 10,000  bars in the back test block we can say with certainty that stripping the last row will be wrong 99.999% of the time as the first 9999 bars were complete. 

We can also say if the end of backtest is in the past the final 1 candle will be complete. 

This is different to live/dry which is only testing the most current candle

Testing confirms this, we are using the wrong price on any backtest candle from the past.

**Comment by xmatthias** [comment]
if there are 10000 bars in backtest, the 10000th (newst) could be partial, and that's the candle which is removed, before starting the backtest loop. this action is done exactly once (per pair to backtest).

i suspect you can get different results - but that's mainly because the "stoploss at end" triggers and uses the last candle.

**Comment by creslinux** [comment]
Awesome, i'll bell that out with a couple test scenarios to confirm.

**Comment by xmatthias** [comment]
just remove the following line - that should be the only place removing a tick from the dataframe in backtesting: 
https://github.com/freqtrade/freqtrade/blob/3540ba37123a931e692417b2a7716c096533d76d/freqtrade/optimize/backtesting.py#L218

**Comment by creslinux** [comment]
@xmatthias  you are correct. My apologies the bad bug 
This impacts ONLY buys/Sales if on the last candle. 

<img width="1133" alt="screen shot 2018-07-10 at 6 10 08 pm" src="https://user-images.githubusercontent.com/34645187/42528938-a43d65d8-846c-11e8-9e53-0d96c082e8b8.png">

**Comment by creslinux** [comment]
I'll roll my test window into the future a few bars to belt and brace ensure last trade would be closed correctly if not on last bar.

**Comment by creslinux** [comment]
Extending to include more recent candles removed the diff seen earlier.
By chance as im using a very lose strategy to hit trades, triggers another - again the last candle 

Im with you on your analysis - this will impact only trades if they're trigger on the last backtest candle. 
A bug but a lower priorty. 

<img width="1175" alt="screen shot 2018-07-10 at 6 19 11 pm" src="https://user-images.githubusercontent.com/34645187/42529438-04a41c7c-846e-11e8-8d50-4d751c958b64.png">

**Comment by creslinux** [comment]
for completeness - testing by hashing out and not to generate the same trades/periods
`frame.drop(frame.tail(1).index, inplace=True)     # eliminate partial candle`

**Comment by creslinux** [comment]
This can be closed, last candle buy is another issue, if issue, in back test than this.

**Comment by creslinux** [comment]
This can be closed, last candle buy is another issue, if issue, in back test than this.
